### PR TITLE
Watershedline

### PR DESCRIPTION
**Pull request by ThomasWalter**
## Description
Added a functionality to the watershed function. The watershed function implemented in scikit-image allows to partition the image plane into regions (according to the watershed algorithm). These regions touch each other. An alternative and widely used variant of the watershed algorithm separates the regions by a one-pixel-wide line. This option was added to the function. 

## Checklist
[It's fine to submit PRs which are a work in progress! But before they are merged, all PRs should provide:]
- [x] Clean style in [the spirit of PEP8](https://www.python.org/dev/peps/pep-0008/)
- [x] [Docstrings for all functions](https://github.com/numpy/numpy/blob/master/doc/example.py)
- [ ] Gallery example in `./doc/examples` (new features only)
- [x] Unit tests

[For detailed information on these and other aspects see [scikit-image contribution guidelines](http://scikit-image.org/docs/dev/contribute.html)]

## References
[If this is a bug-fix or enhancement, it closes issue # ]
[If this is a new feature, it implements the following paper: ]

### CODE REVIEW

**Comment by ThomasWalter** [comment]
Sorry, there is a bug for negative labels ... I am looking at this now. I will tell you when I am done.

**Comment by ThomasWalter** [comment]
OK, this is corrected now.

**Comment by ThomasWalter** [comment]
By the way, I had one question about the implementation. Does the implementation as raveled arrays bring a huge advantage in processing time or is this marginal?

**Comment by ThomasWalter** [comment]
Just wanted to ask: is there anything I can do at the moment to fix the failed check?

**Comment by jni** [comment]
@ThomasWalter the ravelled implementation is primarily for the purpose of supporting 2D, 3D, and n-D algorithms with a single code base.

**Comment by jni** [comment]
@ThomasWalter I don't have time to do a full review right now but in principle I like this addition. In the meantime please go over the diff and fix:

- a remaining stray pdb import
- PEP8 inconsistencies
- trailing white space on some lines

Thanks!

**Comment by ThomasWalter** [comment]
Ok thanks a lot for your comments. 

I guess, that the issues should be solved (unless that there are additional PEP8 inconsistencies I am not aware of, which is perfectly possible). 

I assume that the review will take place after the new release (which is perfectly fine for me). 

I wanted to add some morphological operators next week. Should I continue on this branch or do I need to start from the master branch as usual?

File: skimage/morphology/tests/test_watershed.py [hunk 1]
```diff
@@ -423,6 +424,31 @@ def test_watershed11(self):
...
+                         [ 77,  51,  51, 102, 153, 153, 203, 255, 203, 203, 203, 153, 102, 102, 102, 153],
+                         [ 77,   0,  51, 102, 153, 203, 203, 255, 203, 255, 203, 153, 102,  51, 102, 153],
+                         [ 77,   0,  51, 102, 153, 203, 255, 255, 203, 203, 203, 153, 102,   0, 102, 153],
+                         [102,   0,  51, 102, 153, 203, 255, 203, 203, 153, 153, 153, 102, 102, 102, 153],
+                         [102, 102, 102, 102, 153, 203, 255, 203, 153, 153, 153, 153, 153, 153, 153, 153]])
+        markerbin = data==0
+        marker = label(markerbin)
+        ws = watershed(data, marker, connectivity=2, watershedline=True)
+        for i, res in zip(range(4), [34,74,74,74]):
+            self.assertTrue(np.sum(ws==i)==res)
```
Comment (jni) [inline skimage/morphology/tests/test_watershed.py]: PEP8 — spacing around `==`

File: skimage/morphology/watershed.py [hunk 2]
```diff
@@ -132,7 +131,7 @@ def _compute_neighbors(image, structure, offset):
 
 
 def watershed(image, markers, connectivity=1, offset=None, mask=None,
-              compactness=0):
+              compactness=0, watershedline=False):
```
Comment (jni) [inline skimage/morphology/watershed.py]: I'd prefer `watershed_line`, and I think it's what PEP8 recommends.

File: skimage/morphology/watershed.py [hunk 3]
```diff
@@ -142,7 +141,8 @@ def watershed(image, markers, connectivity=1, offset=None, mask=None,
         Data array where the lowest value points are labeled first.
     markers: int, or ndarray of int, same shape as `image`
         The desired number of markers, or an array marking the basins with the
-        values to be assigned in the label matrix. Zero means not a marker.
+        values to be assigned in the label matrix. Zero means not a marker. The 
+        marker labels are supposed to be positive.
```
Comment (jni) [inline skimage/morphology/watershed.py]: are supposed to -> should be

File: skimage/morphology/watershed.py [hunk 4]
```diff
@@ -255,8 +258,13 @@ def watershed(image, markers, connectivity=1, offset=None, mask=None,
...
                                  mask, image_strides, compactness,
-                                 output.ravel())
+                                 output.ravel(),
+                                 watershedline)
 
     output = crop(output, pad_width, copy=True)
 
+    if watershedline:
+        min_val = output.min()
+        output[output==min_val] = 0
```
Comment (jni) [inline skimage/morphology/watershed.py]: space around `==`.

File: skimage/morphology/tests/test_watershed.py [hunk 5]
```diff
@@ -423,6 +424,31 @@ def test_watershed11(self):
...
+                         [102,  51,  51, 102, 102, 153, 203, 255, 203, 203, 153, 153, 153, 153, 102, 153],
+                         [ 77,  51,  51, 102, 153, 153, 203, 255, 203, 203, 203, 153, 102, 102, 102, 153],
+                         [ 77,   0,  51, 102, 153, 203, 203, 255, 203, 255, 203, 153, 102,  51, 102, 153],
+                         [ 77,   0,  51, 102, 153, 203, 255, 255, 203, 203, 203, 153, 102,   0, 102, 153],
+                         [102,   0,  51, 102, 153, 203, 255, 203, 203, 153, 153, 153, 102, 102, 102, 153],
+                         [102, 102, 102, 102, 153, 203, 255, 203, 153, 153, 153, 153, 153, 153, 153, 153]])
+        markerbin = data==0
+        marker = label(markerbin)
+        ws = watershed(data, marker, connectivity=2, watershedline=True)
+        for i, res in zip(range(4), [34,74,74,74]):
```
Comment (ThomasWalter) [inline skimage/morphology/tests/test_watershed.py]: I agree in principle, but label is already used as a function name. Used `lab` instead.

File: skimage/morphology/watershed.py (hunk 3 above)
Comment (ThomasWalter) [inline skimage/morphology/watershed.py]: Actually, they can be negative (although this is highly unusual) in the last version I pushed. I remove the sentence.

**Comment by ThomasWalter** [comment]
Thanks a lot for your comments. I have modified the code accordingly. 

Regarding the feature freeze: is there any impact on contributing to skimage (other than actually contributing to 0.14)? I mean: is there anything particular I would need to do to make a new contribution?

**Comment by ThomasWalter** [comment]
@Juan OK, great. Thanks a lot. And a happy new year.

File: skimage/morphology/_watershed.pyx [hunk 6]
```diff
@@ -10,6 +10,8 @@ All rights reserved.
 Original author: Lee Kamentsky
 """
 import numpy as np
+from libc.math cimport sqrt
```
Comment (soupault) [inline skimage/morphology/_watershed.pyx]: Why do you need this import?

File: skimage/morphology/tests/test_watershed.py [hunk 7]
```diff
@@ -321,7 +322,7 @@ def test_watershed07(self):
         #
         # The two objects should be the same size, except possibly for the
         # border region
-        #
+#
```
Comment (soupault) [inline skimage/morphology/tests/test_watershed.py]: Revert this, please.

File: skimage/morphology/watershed.py [hunk 8]
```diff
@@ -156,6 +156,9 @@ def watershed(image, markers, connectivity=1, offset=None, mask=None,
     compactness : float, optional
         Use compact watershed [3]_ with given compactness parameter.
         Higher values result in more regularly-shaped watershed basins.
+    watershed_line: bool, optional (default=False)
```
Comment (soupault) [inline skimage/morphology/watershed.py]: Please, remove `(default=False)`. We mainly use `%param%: %type%[, optional]` all over the package.

File: skimage/morphology/tests/test_watershed.py [hunk 9]
```diff
@@ -423,6 +424,31 @@ def test_watershed11(self):
         dmin = np.min(d, 2)
         self.assertTrue(np.all(d[i, j, out[i, j]-1] == dmin))
 
+    def test_watershed12(self):
```
Comment (soupault) [inline skimage/morphology/tests/test_watershed.py]: Use 2 blank lines between functions.

**Comment by soupault** [comment]
| W/o watershed line  | W/  |
|:-:|:-:|
| image | image |

File: skimage/morphology/_watershed.pyx (hunk 6 above)
Comment (ThomasWalter) [inline skimage/morphology/_watershed.pyx]: I personally did not use this sqrt, but some contributor needed this to calculate the euclidean distance in line 41: 

https://github.com/ThomasWalter/scikit-image/blob/66743dc3c344b0f3c8bc60fd45d49ea465fe0784/skimage/morphology/_watershed.pyx#L41

I changed the place of the import for PEP8 reasons.

Comment (soupault) [inline skimage/morphology/_watershed.pyx]: Oh... Of course, sorry, I missed that.

File: skimage/morphology/watershed.py (hunk 8 above)
Comment (ThomasWalter) [inline skimage/morphology/watershed.py]: OK. 

So, you do not write the default value somewhere in the doc?

Comment (soupault) [inline skimage/morphology/watershed.py]: Only in the signature. And, sometimes, in the description if it's not obvious what it means.

P.S. There is a discussion on changing this convention (https://github.com/scikit-image/scikit-image/issues/2409), but without any agreement atm.

Comment (ThomasWalter) [inline skimage/morphology/watershed.py]: OK thanks.

**Comment by soupault** [comment]
Thanks a lot @ThomasWalter !

**Pull request merged**